contracts/payment/MCPayment.sol

```diff
@@ -0,0 +1,172 @@
+    error InvalidSignature(string message);
+    error PaymentError(string message);
+    error WithdrawError(string message);
+
+    /**
+     * @dev Initialize the contract
+     */
+    function initialize(address owner, uint8 ownerPercentage) public initializer {
+        MCPaymentStorage storage $ = _getMCPaymentStorage();
+        $.ownerPercentage = ownerPercentage;
+        __EIP712_init("MCPayment", VERSION);
+        __Ownable_init(owner);
+    }
+
+    function updateOwnerPercentage(uint8 ownerPercentage) external onlyOwner {
+        MCPaymentStorage storage $ = _getMCPaymentStorage();
+        $.ownerPercentage = ownerPercentage;
+    }
+
+    function pay(
+        Iden3PaymentRailsRequestV1 memory paymentData,
+        bytes memory signature
+    ) external payable {
+        verifySignature(paymentData, signature);
+        bytes32 paymentId = keccak256(abi.encode(paymentData.recipient, paymentData.nonce));
+        MCPaymentStorage storage $ = _getMCPaymentStorage();
+        if ($.isPaid[paymentId]) {
+            revert PaymentError("MCPayment: payment already paid");
+        }
+        if (paymentData.amount != msg.value) {
+            revert PaymentError("MCPayment: invalid payment value");
+        }
+        if (paymentData.expirationDate < block.timestamp) {
+            revert PaymentError("MCPayment: payment expired");
+        }
+
+        uint256 ownerPart = (msg.value * $.ownerPercentage) / 100;
+        uint256 issuerPart = msg.value - ownerPart;
+
+        $.balance[paymentData.recipient] += issuerPart;
+        $.ownerBalance += ownerPart;
+
+        emit Payment(paymentData.recipient, paymentData.nonce);
+        $.isPaid[paymentId] = true;
+    }
+
+    function isPaymentDone(address recipient, uint256 nonce) external view returns (bool) {
+        MCPaymentStorage storage $ = _getMCPaymentStorage();
+        return $.isPaid[keccak256(abi.encode(recipient, nonce))];
+    }
+
+    function verifySignature(
+        Iden3PaymentRailsRequestV1 memory paymentData,
+        bytes memory signature
+    ) public view {
+        bytes32 structHash = keccak256(
+            abi.encode(
+                PAYMENT_DATA_TYPE_HASH,
+                paymentData.recipient,
+                paymentData.amount,
+                paymentData.expirationDate,
+                paymentData.nonce,
+                keccak256(paymentData.metadata)
+            )
+        );
+        bytes32 hashTypedData = _hashTypedDataV4(structHash);
+        (address recovered, ECDSA.RecoverError err, ) = hashTypedData.tryRecover(signature);
+
+        if (err != ECDSA.RecoverError.NoError || recovered != paymentData.recipient) {
+            revert InvalidSignature("MCPayment: invalid signature");
+        }
+    }
+
+    function getMyBalance() public view returns (uint256) {
+        MCPaymentStorage storage $ = _getMCPaymentStorage();
+        return $.balance[_msgSender()];
+    }
```

Comment:
Does it really work if you call the method without transaction? The will be no signature so no msg.sender. Maybe `getBalance(address)` is better? Besides, you can ask for any address that way.